store/postgres/src/chain_store.rs

```diff
@@ -908,38 +908,68 @@ mod data {
                 Storage::Shared => {
-                    const ANCESTOR_SQL: &str = "
+                    let query = format!(
+                        "
         with recursive ancestors(block_hash, block_offset) as (
             values ($1, 0)
             union all
             select b.parent_hash, a.block_offset+1
               from ancestors a, ethereum_blocks b
              where a.block_hash = b.hash
                and a.block_offset < $2
+               {}
         )
-        select a.block_hash as hash
+        select a.block_hash as hash, b.number as number
           from ancestors a
-         where a.block_offset = $2;";
+               inner join ethereum_blocks b on a.block_hash = b.hash
+         order by a.block_offset desc limit 1",
+                        short_circuit_predicate
+                    );
+                    // type Result = (Text, i64);
+                    #[derive(QueryableByName)]
+                    struct BlockHashAndNumber {
+                        #[sql_type = "Text"]
+                        hash: String,
+                        #[sql_type = "BigInt"]
+                        number: i64,
+                    }
 
-                    let hash = sql_query(ANCESTOR_SQL)
-                        .bind::<Text, _>(block_ptr.hash_hex())
-                        .bind::<BigInt, _>(offset as i64)
-                        .get_result::<BlockHashText>(conn)
-                        .optional()?;
+                    let block = match root {
+                        Some(root) => sql_query(query)
+                            .bind::<Text, _>(block_ptr.hash_hex())
+                            .bind::<BigInt, _>(offset as i64)
+                            .bind::<Text, _>(root.hash_hex())
+                            .get_result::<BlockHashAndNumber>(conn),
+                        None => sql_query(query)
+                            .bind::<Text, _>(block_ptr.hash_hex())
+                            .bind::<BigInt, _>(offset as i64)
+                            .get_result::<BlockHashAndNumber>(conn),
+                    }
+                    .optional()?;
 
                     use public::ethereum_blocks as b;
 
-                    match hash {
+                    match block {
                         None => None,
-                        Some(hash) => Some((
+                        Some(block) => Some((
                             b::table
-                                .filter(b::hash.eq(&hash.hash))
+                                .filter(b::hash.eq(&block.hash))
                                 .select(b::data)
                                 .first::<json::Value>(conn)?,
-                            BlockHash::from_str(&hash.hash)?,
+                            BlockPtr::new(
+                                BlockHash::from_str(&block.hash)?,
+                                i32::try_from(block.number).unwrap(),
+                            ),
```

Comment:
@lutter could you take a look at this query generation part, looks good to me, but with the recursive cte's just wanted to double check if everythings alright.